pkg/testutils/kubernetes/builders/client.go

```diff
@@ -0,0 +1,212 @@
+package builders
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+const (
+	// ObjectEventAll subscribe to all object events
+	ObjectEventAll ObjectEvent = "ALL"
+	// ObjectEventAdded subscribe to object creation events
+	ObjectEventAdded ObjectEvent = "ADDED"
+	// ObjectEventDeleted subscribe to object delete events
+	ObjectEventDeleted ObjectEvent = "DELETED"
+	// ObjectEventModified subscribe to object update events
+	ObjectEventModified ObjectEvent = "MODIFIED"
+)
+
+// ObjectEvent defines an event in an object
+type ObjectEvent string
+
+// PodObserver is a function that receives notifications of events on an Pod
+// and can update it by returning a non-nil value. In addition, the PodObserver
+// returns a boolean value indicating if it wants to keep receiving events or not.
+//
+// Note: PodObserver that subscribe to update events should implement a mechanism for
+// avoiding an update loop. They can for instance check the object is in a particular state
+// before updating. Additionally, they can unsubscribe from further updates.
+type PodObserver func(ObjectEvent, *corev1.Pod) (*corev1.Pod, bool, error)
```

Comment:
I'm not 100% sold on the name `observer`, I think it has read-only connotations for me. As the concept is somewhat similar to mutation webhooks, what about using `mutator`?

Reply:
I also but `mutator` has the problem as mutating the object is optional. What about `PodEventHook`? It's more neutral.